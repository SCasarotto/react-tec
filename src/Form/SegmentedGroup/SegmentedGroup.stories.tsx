import React from 'react';

import { Story, Meta } from '@storybook/react';

import { SegmentedGroup, SegmentedGroupProps } from './SegmentedGroup';

export default {
  title: 'Form/SegmentedGroup',
  component: SegmentedGroup,
  argTypes: {},
} as Meta;

const Template: Story<SegmentedGroupProps> = (args) => (
  <SegmentedGroup {...args} />
);

export const Default = Template.bind({});
Default.args = {
  labelForKey: 'SegmentedGroup',
  title: 'My Radio Button Group',
  onChange: (e) => console.log(e.target.value),
  buttonArray: ['value1', 'valueTwo', 'OtherValue'],
};

export const Inline = Template.bind({});
Inline.args = {
  labelForKey: 'SegmentedGroupInline',
  title: 'My Radio Button Group',
  onChange: (e) => console.log(e.target.value),
  buttonArray: ['value1', 'valueTwo', 'OtherValue'],
  inline: true,
};
