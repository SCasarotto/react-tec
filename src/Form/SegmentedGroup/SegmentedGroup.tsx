import React from 'react';

import { manipulateRowData } from '../../helpers';
import { FormRowCustomProps } from '../FormRow';
import { LabelCustomProps } from '../Label';
import { RadioButtonInputProps } from '../RadioButtonInput';
import {
  Row,
  Label,
  SegmentedContainer,
  LabelWrapper,
  Input,
  InputLabel,
} from './styledComponents';

export interface SegmentedGroupButtonData {
  label: string;
  value?: string;
  key?: string;
}
export interface SegmentedGroupProps
  extends FormRowCustomProps,
    LabelCustomProps,
    RadioButtonInputProps {
  checkedValue: string;
  buttonArray: Array<string | SegmentedGroupButtonData>;
  inline?: boolean;
  labelForKey: string;
}
/**
 *
 * SegmentedGroup is a stylized `RadioButtonGroup`. This component mimics the style of iOS's Segmented Controls.
 *
 */
export const SegmentedGroup: React.FC<SegmentedGroupProps> = (props) => {
  const {
    title,
    onChange,
    checkedValue,
    buttonArray,
    labelForKey,
    required,
    inline,
    disabled,
    rowSize,
    last,
    className = '',
  } = props;

  return (
    <Row
      rowSize={rowSize}
      last={last}
      inline={inline}
      className={`tec-segmentedGroup ${className}`}
    >
      {title && (
        <Label
          inline={inline}
          required={required}
          disabled={disabled}
          className='tec-segmentedGroupTitle'
        >
          {title}
        </Label>
      )}
      <SegmentedContainer
        inline={inline}
        className='tec-segmentedGroupSegmentedContainer'
      >
        {buttonArray?.map((rowData, index) => {
          const { label, value, key } = manipulateRowData({
            rowData,
            labelForKey,
          });

          return (
            <LabelWrapper
              key={key}
              className='tec-segmentedGroupSegmentedLabelWrapper'
            >
              <Input
                value={value}
                onChange={onChange}
                id={key}
                checked={checkedValue === value}
                disabled={disabled}
                className='tec-segmentedGroupSegmentedInput'
              />
              <InputLabel
                htmlFor={key}
                key={key}
                first={index === 0}
                last={index + 1 === buttonArray.length}
                checked={checkedValue === value}
                disabled={disabled}
                className='tec-segmentedGroupSegmentedLabel'
              >
                {label}
              </InputLabel>
            </LabelWrapper>
          );
        })}
      </SegmentedContainer>
    </Row>
  );
};
