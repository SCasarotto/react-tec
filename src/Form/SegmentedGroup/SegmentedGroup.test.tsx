import React from 'react';

import { render } from '@testing-library/react';

import { SegmentedGroup } from './SegmentedGroup';

describe('SegmentedGroup', () => {
  // Renders Without Error
  it('matches snapshot', () => {
    const { asFragment } = render(
      <SegmentedGroup
        labelForKey='testSegmentedGroup'
        checkedValue='someValue'
        buttonArray={['someValue', 'anotherValue']}
        onChange={jest.fn()}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  // Class Name
  it('accepts custom className', () => {
    const { container } = render(
      <SegmentedGroup
        labelForKey='testSegmentedGroup'
        checkedValue='someValue'
        buttonArray={['someValue', 'anotherValue']}
        onChange={jest.fn()}
        className='myClassName'
      />,
    );
    expect(container.firstChild).toHaveClass('myClassName');
  });
});
