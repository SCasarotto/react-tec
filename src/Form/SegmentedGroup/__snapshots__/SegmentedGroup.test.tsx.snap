// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentedGroup matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  display: inline-block;
  vertical-align: top;
  margin-bottom: 10px;
  text-align: left;
  box-sizing: border-box;
  padding-right: 0px;
  width: 100%;
}

.c4 {
  position: relative;
  display: block;
  width: auto;
  box-sizing: border-box;
  margin-bottom: 3px;
  font-size: 18px;
  cursor: pointer;
}

.c2 {
  cursor: pointer;
  box-sizing: border-box;
  margin: 0px;
}

.c2:disabled {
  cursor: not-allowed;
}

.c5 {
  width: 100%;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
  text-align: left;
  display: block;
  margin-bottom: 3px;
  padding-right: 0px;
  font-size: 18px;
}

.c1 {
  display: inline-block;
  width: auto;
}

.c3 {
  display: none;
}

.c6 {
  display: inline-block;
  width: auto;
  border: 1px solid;
  padding: 8px 15px;
  margin: 0px;
  font-size: 14px;
  cursor: pointer;
  pointer-events: auto;
  -webkit-transition: color 0.2s ease-in,background-color 0.2s ease-in, border 0.2s ease-in;
  transition: color 0.2s ease-in,background-color 0.2s ease-in, border 0.2s ease-in;
  border-top-left-radius: 5px;
  border-bottom-left-radius: 5px;
}

.c6:hover,
.c6:active {
  border: 1px solid;
}

.c7 {
  display: inline-block;
  width: auto;
  border: 1px solid;
  padding: 8px 15px;
  margin: 0px;
  font-size: 14px;
  cursor: pointer;
  pointer-events: auto;
  -webkit-transition: color 0.2s ease-in,background-color 0.2s ease-in, border 0.2s ease-in;
  transition: color 0.2s ease-in,background-color 0.2s ease-in, border 0.2s ease-in;
  border-top-right-radius: 5px;
  border-bottom-right-radius: 5px;
}

.c7:hover,
.c7:active {
  border: 1px solid;
}

@media (max-width:550px) {
  .c4 {
    font-size: 16px;
  }
}

@media (max-width:550px) {
  .c5 {
    font-size: 16px;
  }
}

<div
    class="c0 tec-formRow tec-segmentedGroup "
  >
    <div
      class="tec-segmentedGroupSegmentedContainer"
    >
      <div
        class="c1 tec-segmentedGroupSegmentedLabelWrapper"
      >
        <input
          checked=""
          class="c2 tec-radioButtonInput c3 tec-segmentedGroupSegmentedInput"
          id="testSegmentedGroupsomeValue"
          type="radio"
          value="someValue"
        />
        <label
          class="c4 tec-label c5 c6 tec-segmentedGroupSegmentedLabel"
          for="testSegmentedGroupsomeValue"
        >
          someValue
        </label>
      </div>
      <div
        class="c1 tec-segmentedGroupSegmentedLabelWrapper"
      >
        <input
          class="c2 tec-radioButtonInput c3 tec-segmentedGroupSegmentedInput"
          id="testSegmentedGroupanotherValue"
          type="radio"
          value="anotherValue"
        />
        <label
          class="c4 tec-label c5 c7 tec-segmentedGroupSegmentedLabel"
          for="testSegmentedGroupanotherValue"
        >
          anotherValue
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
